In [43]:
import configparser
import psycopg2
import pandas as pd
import boto3
import json

## 1. Read configuration from file

Read from file dwh.cfg and print out in dataframe

In [44]:
config = configparser.ConfigParser()
config.read_file(open('dwh.cfg'))

KEY                    = config.get('AWS','KEY')
SECRET                 = config.get('AWS','SECRET')

DWH_CLUSTER_TYPE       = config.get("DWH","DWH_CLUSTER_TYPE")
DWH_NUM_NODES          = config.get("DWH","DWH_NUM_NODES")
DWH_NODE_TYPE          = config.get("DWH","DWH_NODE_TYPE")

DWH_CLUSTER_IDENTIFIER = config.get("DWH","dwh_cluster_identifier")
DWH_DB                 = config.get("CLUSTER","db_name")
DWH_DB_USER            = config.get("CLUSTER","db_user")
DWH_DB_PASSWORD        = config.get("CLUSTER","db_password")
DWH_PORT               = config.get("CLUSTER","db_port")

DWH_IAM_ROLE_NAME      = config.get("IAM_ROLE", "iam_role_name")

(DWH_DB_USER, DWH_DB_PASSWORD, DWH_DB)

pd.DataFrame({"Param":
                  ["DWH_CLUSTER_TYPE", "DWH_NUM_NODES", "DWH_NODE_TYPE", "DWH_CLUSTER_IDENTIFIER", 
                   "DWH_DB", "DWH_DB_USER", "DWH_DB_PASSWORD", "DWH_PORT", "DWH_IAM_ROLE_NAME"],
              "Value":
                  [DWH_CLUSTER_TYPE, DWH_NUM_NODES, DWH_NODE_TYPE, DWH_CLUSTER_IDENTIFIER, DWH_DB, 
                   DWH_DB_USER, DWH_DB_PASSWORD, DWH_PORT, DWH_IAM_ROLE_NAME]
             })

Param       Value
0  DWH_CLUSTER_TYPE        multi-node
1  DWH_NUM_NODES           4         
2  DWH_NODE_TYPE           dc2.large 
3  DWH_CLUSTER_IDENTIFIER  dwhCluster
4  DWH_DB                  sparkify  
5  DWH_DB_USER             dwhuser   
6  DWH_DB_PASSWORD         Passw0rd  
7  DWH_PORT                5439      
8  DWH_IAM_ROLE_NAME       dwhRole

## 2. Create clients for EC2, S3, IAM, and Redshift

In [45]:
ec2 = boto3.resource('ec2',
                       region_name="us-west-2",
                       aws_access_key_id=KEY,
                       aws_secret_access_key=SECRET
                    )

s3 = boto3.resource('s3',
                       region_name="us-west-2",
                       aws_access_key_id=KEY,
                       aws_secret_access_key=SECRET
                   )

iam = boto3.client('iam',aws_access_key_id=KEY,
                     aws_secret_access_key=SECRET,
                     region_name='us-west-2'
                  )

redshift = boto3.client('redshift',
                       region_name="us-west-2",
                       aws_access_key_id=KEY,
                       aws_secret_access_key=SECRET
                       )

### Print IAM  role ARN 

In [46]:
print('IAM role ARN : ')
roleArn = iam.get_role(RoleName=DWH_IAM_ROLE_NAME)['Role']['Arn']

print(roleArn)

IAM role ARN : 
arn:aws:iam::497483064811:role/dwhRole


## 3. Create Redshift Cluster

#### To configure cluster we add following configuration: 
- hardware parameters
- identifiers and credentials parameters
- role parametes

<span style="color:red">! Below block will create cluster</span>

In [47]:
try:
    response = redshift.create_cluster( 
# Adding parameters for hardware:
        ClusterType=DWH_CLUSTER_TYPE,
        NodeType=DWH_NODE_TYPE,
        NumberOfNodes=int(DWH_NUM_NODES),
       

# Add parameters for identifiers & credentials
        DBName=DWH_DB,
        ClusterIdentifier=DWH_CLUSTER_IDENTIFIER,
        MasterUsername=DWH_DB_USER,
        MasterUserPassword=DWH_DB_PASSWORD,
        
# Add parameter for role (to allow s3 access)
        IamRoles=[roleArn]  
    )
except Exception as e:
    print(e)

### 3.1 Check the cluster status 

<span style="color:blue">run this block several times until the cluster status becomes Available</span>

In [49]:
def prettyRedshiftProps(props):
    pd.set_option('display.max_colwidth', -1)
    keysToShow = ["ClusterIdentifier", "NodeType", "ClusterStatus", "MasterUsername", "DBName", 
                  "Endpoint", "NumberOfNodes", 'VpcId']
    x = [(k, v) for k,v in props.items() if k in keysToShow]
    return pd.DataFrame(data=x, columns=["Key", "Value"])

myClusterProps = redshift.describe_clusters(ClusterIdentifier=DWH_CLUSTER_IDENTIFIER)['Clusters'][0]
prettyRedshiftProps(myClusterProps)

Key  \
0  ClusterIdentifier   
1  NodeType            
2  ClusterStatus       
3  MasterUsername      
4  DBName              
5  Endpoint            
6  VpcId               
7  NumberOfNodes       

                                                                                   Value  
0  dwhcluster                                                                             
1  dc2.large                                                                              
2  available                                                                              
3  dwhuser                                                                                
4  sparkify                                                                               
5  {'Address': 'dwhcluster.cxhm3t9hpxvk.us-west-2.redshift.amazonaws.com', 'Port': 5439}  
6  vpc-b1ea40c9                                                                           
7  4

### 3.2 Retreive clusters ENDPOINT and ROLE_ARN

<span style="color:red">! Do not run before cluster is available </span> 

DWH_ENDPOINT ::  dwhcluster.cxhm3t9hpxvk.us-west-2.redshift.amazonaws.com

DWH_ROLE_ARN ::  arn:aws:iam::497483064811:role/dwhRole

In [50]:
DWH_ENDPOINT = myClusterProps['Endpoint']['Address']
DWH_ROLE_ARN = myClusterProps['IamRoles'][0]['IamRoleArn']
print("DWH_ENDPOINT :: ", DWH_ENDPOINT)
print("DWH_ROLE_ARN :: ", roleArn)

DWH_ENDPOINT ::  dwhcluster.cxhm3t9hpxvk.us-west-2.redshift.amazonaws.com
DWH_ROLE_ARN ::  arn:aws:iam::497483064811:role/dwhRole


### 4. Open TCP connection to cluster

In [51]:
try:
    vpc = ec2.Vpc(id=myClusterProps['VpcId'])
    defaultSg = list(vpc.security_groups.all())[0]
    print(defaultSg)
    
    defaultSg.authorize_ingress(
        GroupName=defaultSg.group_name,
        CidrIp='0.0.0.0/0',
        IpProtocol='TCP', 
        FromPort=int(DWH_PORT),
        ToPort=int(DWH_PORT)
    )
except Exception as e:
    print(e)

ec2.SecurityGroup(id='sg-f9289eb4')
An error occurred (InvalidPermission.Duplicate) when calling the AuthorizeSecurityGroupIngress operation: the specified rule "peer: 0.0.0.0/0, TCP, from port: 5439, to port: 5439, ALLOW" already exists


In [52]:
# Connection sting

conn_string="postgresql://{}:{}@{}:{}/{}".format(DWH_DB_USER, DWH_DB_PASSWORD, DWH_ENDPOINT, DWH_PORT,DWH_DB)
print(conn_string)

postgresql://dwhuser:Passw0rd@dwhcluster.cxhm3t9hpxvk.us-west-2.redshift.amazonaws.com:5439/sparkify


### To be able to use  %sql function, you might need to install ipython-sql first

Run below command:

`` !pip install ipython-sql``

Then load sql and call $conn_string defined above to validate your connection to Redshift DB

In [27]:
# Load SQL extention and check connection string 
%load_ext sql
%sql $conn_string

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: dwhuser@sparkify'

In [53]:
# Test connection to DB 
%%sql
SELECT * FROM staging_events
LIMIT  10;

 * postgresql://dwhuser:***@dwhcluster.cxhm3t9hpxvk.us-west-2.redshift.amazonaws.com:5439/sparkify
(psycopg2.OperationalError) SSL SYSCALL error: Operation timed out
 [SQL: 'SELECT * FROM staging_events\nLIMIT  10;'] (Background on this error at: http://sqlalche.me/e/e3q8)


## 5. Clean up your resources

<span style="color:red"> CAREFUL!!</span>

Below command will delete your cluster


In [54]:
#### CAREFUL!!
#-- Uncomment & run to delete the created resources
redshift.delete_cluster( ClusterIdentifier=DWH_CLUSTER_IDENTIFIER,  SkipFinalClusterSnapshot=True)


{'Cluster': {'ClusterIdentifier': 'dwhcluster',
  'NodeType': 'dc2.large',
  'ClusterStatus': 'deleting',
  'MasterUsername': 'dwhuser',
  'DBName': 'sparkify',
  'Endpoint': {'Address': 'dwhcluster.cxhm3t9hpxvk.us-west-2.redshift.amazonaws.com',
   'Port': 5439},
  'ClusterCreateTime': datetime.datetime(2019, 6, 3, 20, 48, 13, 707000, tzinfo=tzutc()),
  'AutomatedSnapshotRetentionPeriod': 1,
  'ManualSnapshotRetentionPeriod': -1,
  'ClusterSecurityGroups': [],
  'VpcSecurityGroups': [],
  'ClusterParameterGroups': [{'ParameterGroupName': 'default.redshift-1.0',
    'ParameterApplyStatus': 'in-sync'}],
  'ClusterSubnetGroupName': 'default',
  'VpcId': 'vpc-b1ea40c9',
  'AvailabilityZone': 'us-west-2d',
  'PreferredMaintenanceWindow': 'thu:11:00-thu:11:30',
  'PendingModifiedValues': {},
  'ClusterVersion': '1.0',
  'AllowVersionUpgrade': True,
  'NumberOfNodes': 4,
  'PubliclyAccessible': True,
  'Encrypted': False,
  'Tags': [],
  'EnhancedVpcRouting': False,
  'IamRoles': [{'IamRoleA

In [29]:
#### Run this cell to check status

myClusterProps = redshift.describe_clusters(ClusterIdentifier=DWH_CLUSTER_IDENTIFIER)['Clusters'][0]
prettyRedshiftProps(myClusterProps)

Key  \
0  ClusterIdentifier   
1  NodeType            
2  ClusterStatus       
3  MasterUsername      
4  DBName              
5  Endpoint            
6  VpcId               
7  NumberOfNodes       

                                                                                   Value  
0  dwhcluster                                                                             
1  dc2.large                                                                              
2  deleting                                                                               
3  dwhuser                                                                                
4  sparkify                                                                               
5  {'Address': 'dwhcluster.cxhm3t9hpxvk.us-west-2.redshift.amazonaws.com', 'Port': 5439}  
6  vpc-b1ea40c9                                                                           
7  4